Address static analyzer warnings and UNSAFE_BUFFER_USAGE in ParkingLot

geoffreygaren opened the pull request:
#### b436f5fd9c2e60d9ee82a047eb67a73e74129da2
<pre>
Address static analyzer warnings and UNSAFE_BUFFER_USAGE in ParkingLot
<a href="https://bugs.webkit.org/show_bug.cgi?id=287446">https://bugs.webkit.org/show_bug.cgi?id=287446</a>
&lt;<a href="https://rdar.apple.com/problem/144574023">rdar://problem/144574023</a>&gt;

Reviewed by Chris Dumez.

Use FixedVector in Hashtable to avoid unsafe pointer arithmetic.

Use RefPtr instead of raw pointer for ThreadData to avoid ambiguous lifetime.
In theory all these RefPtrs are not needed, since ThreadData is per-thread
and a thread will not exit while holding a lock, but it&apos;s nice not to have to
theorize, and also pthread_exit() is a function and per-thread destruction order
is not guaranteed.

* Source/WTF/wtf/ParkingLot.cpp:
(WTF::ParkingLot::parkConditionallyImpl):
(WTF::ParkingLot::forEachImpl):

Canonical link: <a href="https://commits.webkit.org/290217@main">https://commits.webkit.org/290217@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47ad9c9b06fc90891a1d8daaeeb87d1b602555a4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89333 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8858 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44172 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94319 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40094 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17103 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68823 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26489 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92335 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/7100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/81078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49183 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/6846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39201 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/82132 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/36449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96148 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/88109 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/16513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12130 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77697 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/16769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/76868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76996 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9663 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14000 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/16527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/110602 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/16268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26523 "Found 3 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.no-llint, microbenchmarks/memcpy-wasm-small.js.dfg-eager, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-collect-continuously (failure)") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/19719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/18049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -67,7 +66,7 @@ struct ThreadData : public ThreadSafeRefCounted<ThreadData> {
     const void* address { nullptr };
     
-    ThreadData* nextInQueue { nullptr };
+    RefPtr<ThreadData> nextInQueue { nullptr };
```
` { nullptr }` is probably no longer necessary.

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -138,8 +137,8 @@ struct Bucket {
-        ThreadData** currentPtr = &queueHead;
-        ThreadData* previous = nullptr;
+        RefPtr<ThreadData>* currentPtr = &queueHead;
+        RefPtr<ThreadData> previous = nullptr;
```
` = nullptr` is unnecessary.

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -194,8 +193,8 @@ struct Bucket {
 
-    ThreadData* queueHead { nullptr };
-    ThreadData* queueTail { nullptr };
+    RefPtr<ThreadData> queueHead { nullptr };
```
` { nullptr }` unnecessary.

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -194,8 +193,8 @@ struct Bucket {
-    ThreadData* queueHead { nullptr };
-    ThreadData* queueTail { nullptr };
+    RefPtr<ThreadData> queueHead { nullptr };
+    RefPtr<ThreadData> queueTail { nullptr };
```
Ditto.

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -276,14 +269,12 @@ Hashtable* ensureHashtable()
-                    dataLogForCurrentThread(": created initial hashtable ", RawPointer(currentHashtable), "\n");
-                return currentHashtable;
+                    dataLogForCurrentThread(": created initial hashtable ", RawPointer(currentHashtable.get()), "\n");
+                return currentHashtable.release(); // Leak the hash table
```
Comments should end with a period.

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -376,23 +367,23 @@ void ensureHashtableSize(unsigned numThreads)
     for (Bucket* bucket : reusableBuckets) {
-        while (ThreadData* threadData = bucket->dequeue())
-            threadDatas.append(threadData);
+        while (RefPtr<ThreadData> threadData = bucket->dequeue())
```
We probably don't need the `<ThreadData>`

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -376,23 +367,23 @@ void ensureHashtableSize(unsigned numThreads)
-        dataLogForCurrentThread(": created new hashtable: ", RawPointer(newHashtable), "\n");
-    for (ThreadData* threadData : threadDatas) {
+        dataLogForCurrentThread(": created new hashtable: ", RawPointer(newHashtable.get()), "\n");
+    for (auto threadData : threadDatas) {
```
Couldn't we use `auto&`? Or would static analysis complain?

geoffreygaren replied:
I was afraid it would complain but I'll give it a shot.

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -425,9 +416,11 @@ void ensureHashtableSize(unsigned numThreads)
+    bool result = hashtable.compareExchangeStrong(oldHashtable, newHashtable.get()) == oldHashtable;
     RELEASE_ASSERT(result);
 
+    newHashtable.release(); // Leak the hash table
```
Comments should end with a period.

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -425,9 +416,11 @@ void ensureHashtableSize(unsigned numThreads)
     // roll. After we install the new hashtable, we can release all bucket locks.
     
-    bool result = hashtable.compareExchangeStrong(oldHashtable, newHashtable) == oldHashtable;
+    bool result = hashtable.compareExchangeStrong(oldHashtable, newHashtable.get()) == oldHashtable;
```
While don't we use `release()` here instead of `get()`?

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -503,12 +496,12 @@ bool enqueue(const void* address, const Functor& functor)
         }
 
-        ThreadData* threadData = functor();
+        RefPtr<ThreadData> threadData = functor();
```
Do we really need the `<ThreadData>`?

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -578,7 +571,7 @@ NEVER_INLINE ParkingLot::ParkResult ParkingLot::parkConditionallyImpl(
         dataLogForCurrentThread(": parking.\n");
     
-    ThreadData* me = myThreadData();
+    RefPtr<ThreadData> me = myThreadData();
```
ditto.

cdumez commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -806,17 +799,15 @@ NEVER_INLINE void ParkingLot::forEachImpl(const ScopedLambda<void(Thread&, const
         if (!bucket)
             continue;
-        for (ThreadData* currentThreadData = bucket->queueHead; currentThreadData; currentThreadData = currentThreadData->nextInQueue)
+        for (RefPtr<ThreadData> currentThreadData = bucket->queueHead; currentThreadData; currentThreadData = currentThreadData->nextInQueue)
```
Do we really need the `<ThreadData>`?

geoffreygaren committed:
Address static analyzer warnings and UNSAFE_BUFFER_USAGE in ParkingLot
https://bugs.webkit.org/show_bug.cgi?id=287446
<rdar://problem/144574023>

Reviewed by Chris Dumez.

Use FixedVector in Hashtable to avoid unsafe pointer arithmetic.

Use RefPtr instead of raw pointer for ThreadData to avoid ambiguous lifetime.
In theory all these RefPtrs are not needed, since ThreadData is per-thread
and a thread will not exit while holding a lock, but it's nice not to have to
theorize, and also pthread_exit() is a function and per-thread destruction order
is not guaranteed.

* Source/WTF/wtf/ParkingLot.cpp:
(WTF::ParkingLot::parkConditionallyImpl):
(WTF::ParkingLot::forEachImpl):

Canonical link: https://commits.webkit.org/290217@main
